core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -631,29 +641,33 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
       force: Boolean): Seq[String] = {
     logInfo(s"Requesting to kill executor(s) ${executorIds.mkString(", ")}")
 
-    val response = synchronized {
-      val (knownExecutors, unknownExecutors) = executorIds.partition(executorDataMap.contains)
-      unknownExecutors.foreach { id =>
-        logWarning(s"Executor to kill $id does not exist!")
-      }
+    // SPARK-27112: We need to ensure that there is ordering of lock acquisition
+    // between TaskSchedulerImpl and CoarseGrainedSchedulerBackend objects in order to fix
+    // the deadlock issue exposed in SPARK-27112
+    val response = scheduler.synchronized {
+      this.synchronized {
+        val (knownExecutors, unknownExecutors) = executorIds.partition(executorDataMap.contains)
+        unknownExecutors.foreach { id =>
+          logWarning(s"Executor to kill $id does not exist!")
+        }
 
-      // If an executor is already pending to be removed, do not kill it again (SPARK-9795)
-      // If this executor is busy, do not kill it unless we are told to force kill it (SPARK-9552)
-      val executorsToKill = knownExecutors
-        .filter { id => !executorsPendingToRemove.contains(id) }
-        .filter { id => force || !scheduler.isExecutorBusy(id) }
-      executorsToKill.foreach { id => executorsPendingToRemove(id) = !countFailures }
+        // If an executor is already pending to be removed, do not kill it again (SPARK-9795)
+        // If this executor is busy, do not kill it unless we are told to force kill it (SPARK-9552)
+        val executorsToKill = knownExecutors
+          .filter { id => !executorsPendingToRemove.contains(id) }
+          .filter { id => force || !scheduler.isExecutorBusy(id) }
```

Comment:
In a similar vein to my previous comment, although I'm less sure about this one.

This seems to be the only interaction with the scheduler in this method, so could this filtering be done first thing in the method, with the scheduler lock held, and then the rest of the code just needs the `CoarseGrainedSchedulerBackend` lock?

It seems to me the behavior wouldn't change from the current state (where the internal scheduler state can change while this  method is running). And as in the other case, easier to understand things when you're only holding one lock.

Reply:
(Caught up with the previous discussion and it seems that here both locks are needed to avoid an edge case where you could kill active executors.)